In [38]:
import pandas as pd
import numpy as np
import os

In [23]:
df_1 = pd.read_csv(os.path.join("Вариант", "Серия_1", "FMWY.txt"), index_col=0)
df_1 = df_1.rename(columns={'X': 'X_0', 'Y': 'Y_0', "Z": "Z_0"})
df_1.index = pd.to_datetime(df_1.index)
df_1

,X_0,Y_0,Z_0
DATE_TIME,,,
2024-06-13 12:20:38,53030.663033,75477.565580,586.451142
2024-06-13 12:20:53,53030.651604,75477.525852,586.102103
2024-06-13 12:21:08,53030.598681,75477.517472,586.550436
2024-06-13 12:21:23,53030.674503,75477.540323,586.923699
2024-06-13 12:21:38,53030.705023,75477.520573,586.460999
...,...,...,...
2024-06-13 12:47:38,53030.590044,75477.491636,586.368656
2024-06-13 12:47:53,53030.598979,75477.538729,586.113366
2024-06-13 12:48:08,53030.692372,75477.583359,586.020355


In [24]:
df_2 = pd.read_csv(os.path.join("Вариант", "Серия_1", "WLTU.txt"), index_col=0)
df_2 = df_2.rename(columns={'X': 'X_1', 'Y': 'Y_1', "Z": "Z_1"})
df_2.index = pd.to_datetime(df_2.index)
df_2

,X_1,Y_1,Z_1
DATE_TIME,,,
2024-06-13 12:20:23,60791.456970,92126.927878,730.455038
2024-06-13 12:20:38,60791.453622,92126.924734,730.437537
2024-06-13 12:20:53,60791.448412,92126.932076,730.440991
2024-06-13 12:21:08,60791.465849,92126.926478,730.450237
2024-06-13 12:21:23,60791.458409,92126.914855,730.450304
...,...,...,...
2024-06-13 12:47:38,60791.457508,92126.921005,730.440893
2024-06-13 12:48:08,60791.462954,92126.929372,730.471413
2024-06-13 12:48:23,60791.459224,92126.926090,730.434784


In [25]:
df_3 = pd.concat([df_1, df_2], axis=1)
df_3

,X_0,Y_0,Z_0,X_1,Y_1,Z_1
DATE_TIME,,,,,,
2024-06-13 12:20:23,NaN,NaN,NaN,60791.456970,92126.927878,730.455038
2024-06-13 12:20:38,53030.663033,75477.565580,586.451142,60791.453622,92126.924734,730.437537
2024-06-13 12:20:53,53030.651604,75477.525852,586.102103,60791.448412,92126.932076,730.440991
2024-06-13 12:21:08,53030.598681,75477.517472,586.550436,60791.465849,92126.926478,730.450237
2024-06-13 12:21:23,53030.674503,75477.540323,586.923699,60791.458409,92126.914855,730.450304
...,...,...,...,...,...,...
2024-06-13 12:47:53,53030.598979,75477.538729,586.113366,NaN,NaN,NaN
2024-06-13 12:48:08,53030.692372,75477.583359,586.020355,60791.462954,92126.929372,730.471413
2024-06-13 12:48:23,53030.581613,75477.534817,586.116387,60791.459224,92126.926090,730.434784


In [26]:
df_4 = df_3.dropna()
df_4

,X_0,Y_0,Z_0,X_1,Y_1,Z_1
DATE_TIME,,,,,,
2024-06-13 12:20:38,53030.663033,75477.565580,586.451142,60791.453622,92126.924734,730.437537
2024-06-13 12:20:53,53030.651604,75477.525852,586.102103,60791.448412,92126.932076,730.440991
2024-06-13 12:21:08,53030.598681,75477.517472,586.550436,60791.465849,92126.926478,730.450237
2024-06-13 12:21:23,53030.674503,75477.540323,586.923699,60791.458409,92126.914855,730.450304
2024-06-13 12:21:38,53030.705023,75477.520573,586.460999,60791.459772,92126.935256,730.480758
...,...,...,...,...,...,...
2024-06-13 12:47:23,53030.646165,75477.467640,586.110813,60791.462775,92126.935540,730.385742
2024-06-13 12:47:38,53030.590044,75477.491636,586.368656,60791.457508,92126.921005,730.440893
2024-06-13 12:48:08,53030.692372,75477.583359,586.020355,60791.462954,92126.929372,730.471413


In [32]:
df_4.loc[:, 'dX'] = df_4["X_1"] - df_4["X_0"]
df_4.loc[:, 'dY'] = df_4["Y_1"] - df_4["Y_0"] 
df_4.loc[:, 'dZ'] = df_4["Z_1"] - df_4["Z_0"]
df_4

,X_0,Y_0,Z_0,X_1,Y_1,Z_1,dX,dY,dZ
DATE_TIME,,,,,,,,,
2024-06-13 12:20:38,53030.663033,75477.565580,586.451142,60791.453622,92126.924734,730.437537,7760.790589,16649.359154,143.986395
2024-06-13 12:20:53,53030.651604,75477.525852,586.102103,60791.448412,92126.932076,730.440991,7760.796808,16649.406224,144.338888
2024-06-13 12:21:08,53030.598681,75477.517472,586.550436,60791.465849,92126.926478,730.450237,7760.867168,16649.409006,143.899801
2024-06-13 12:21:23,53030.674503,75477.540323,586.923699,60791.458409,92126.914855,730.450304,7760.783906,16649.374532,143.526605
2024-06-13 12:21:38,53030.705023,75477.520573,586.460999,60791.459772,92126.935256,730.480758,7760.754749,16649.414683,144.019759
...,...,...,...,...,...,...,...,...,...
2024-06-13 12:47:23,53030.646165,75477.467640,586.110813,60791.462775,92126.935540,730.385742,7760.816609,16649.467900,144.274929
2024-06-13 12:47:38,53030.590044,75477.491636,586.368656,60791.457508,92126.921005,730.440893,7760.867464,16649.429369,144.072238
2024-06-13 12:48:08,53030.692372,75477.583359,586.020355,60791.462954,92126.929372,730.471413,7760.770582,16649.346013,144.451057


In [29]:
df_5 = df_4[["dX", "dY", "dZ"]]
df_5

,dX,dY,dZ
DATE_TIME,,,
2024-06-13 12:20:38,7760.790589,16649.359154,143.986395
2024-06-13 12:20:53,7760.796808,16649.406224,144.338888
2024-06-13 12:21:08,7760.867168,16649.409006,143.899801
2024-06-13 12:21:23,7760.783906,16649.374532,143.526605
2024-06-13 12:21:38,7760.754749,16649.414683,144.019759
...,...,...,...
2024-06-13 12:47:23,7760.816609,16649.467900,144.274929
2024-06-13 12:47:38,7760.867464,16649.429369,144.072238
2024-06-13 12:48:08,7760.770582,16649.346013,144.451057


In [48]:
df_6 = df_4[["dX", "dY", "dZ"]]
df_6['s_dist'] = (df_6["dX"] ** 2 + df_6["dY"] ** 2 + df_6["dZ"] ** 2) ** 0.5
df_6['azimuth'] = np.degrees(np.arctan2(df_6['dY'], df_6['dX']))
df_6['zenith'] = np.degrees(np.arccos(df_6['dZ'] / df_6['s_dist']))
df_6

/var/folders/c2/pfk652nd5r30mk7t_rz17pg00000gn/T/ipykernel_16488/3560237167.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_6['s_dist'] = (df_6["dX"] ** 2 + df_6["dY"] ** 2 + df_6["dZ"] ** 2) ** 0.5
/var/folders/c2/pfk652nd5r30mk7t_rz17pg00000gn/T/ipykernel_16488/3560237167.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_6['azimuth'] = np.degrees(np.arctan2(df_6['dY'], df_6['dX']))
/var/folders/c2/pfk652nd5r30mk7t_rz17pg00000gn/T/ipykernel_16488/3560237167.py:4: SettingWithCopyWarning: 
A valu

,dX,dY,dZ,s_dist,azimuth,zenith
DATE_TIME,,,,,,
2024-06-13 12:20:38,7760.790589,16649.359154,143.986395,18369.860176,65.008291,89.550900
2024-06-13 12:20:53,7760.796808,16649.406224,144.338888,18369.908231,65.008336,89.549802
2024-06-13 12:21:08,7760.867168,16649.409006,143.899801,18369.937033,65.008141,89.551172
2024-06-13 12:21:23,7760.783906,16649.374532,143.526605,18369.867693,65.008331,89.552335
2024-06-13 12:21:38,7760.754749,16649.414683,144.019759,18369.895625,65.008466,89.550797
...,...,...,...,...,...,...
2024-06-13 12:47:23,7760.816609,16649.467900,144.274929,18369.971994,65.008361,89.550003
2024-06-13 12:47:38,7760.867464,16649.429369,144.072238,18369.956966,65.008167,89.550635
2024-06-13 12:48:08,7760.770582,16649.346013,144.451057,18369.843461,65.008331,89.549451


In [72]:
dsc = df_6.describe()[["s_dist", "azimuth", "zenith"]].loc[["mean", "std"]]
dsc

,s_dist,azimuth,zenith
mean,18369.909069,65.008301,89.550069
std,0.057039,0.000189,0.000927


In [73]:
dsc.loc[["std"]] = dsc.loc[["std"]].apply(lambda row: row / (len(df_6) ** 0.5), axis=1)
dsc

,s_dist,azimuth,zenith
mean,18369.909069,65.008301,89.550069
std,0.006012,0.000020,0.000098


In [74]:
dsc.loc["std", 'azimuth']

1.995042265405436e-05

In [75]:
dsc.loc["std", "azimuth"] = dsc.loc["std", 'azimuth'] * 3600
dsc.loc["std", "zenith"] = dsc.loc["std", "zenith"] * 3600
dsc

,s_dist,azimuth,zenith
mean,18369.909069,65.008301,89.550069
std,0.006012,0.071822,0.351887


GnssVector(FMWY-WLTU, dist=18369.909, azimuth=65.008, zenith=89.550, mse_dx=0.058, mse_dy=0.060, mse_dz=0.297 mse_s_dist=0.006, mse_azimuth=0.072, mse_zenith=0.352)


In [31]:
df_5.loc[:, 's_dist'] = (df_5["dX"] ** 2 + df_5["dY"] ** 2 + df_5["dZ"] ** 2) ** 0.5
df_5

,dX,dY,dZ,s_dist
DATE_TIME,,,,
2024-06-13 12:20:38,7760.790589,16649.359154,143.986395,18369.860176
2024-06-13 12:20:53,7760.796808,16649.406224,144.338888,18369.908231
2024-06-13 12:21:08,7760.867168,16649.409006,143.899801,18369.937033
2024-06-13 12:21:23,7760.783906,16649.374532,143.526605,18369.867693
2024-06-13 12:21:38,7760.754749,16649.414683,144.019759,18369.895625
...,...,...,...,...
2024-06-13 12:47:23,7760.816609,16649.467900,144.274929,18369.971994
2024-06-13 12:47:38,7760.867464,16649.429369,144.072238,18369.956966
2024-06-13 12:48:08,7760.770582,16649.346013,144.451057,18369.843461


In [33]:
mean_s_dist = df_5["s_dist"].mean()
mean_s_dist

18369.90906888889

In [35]:
df_5.loc[:, 'v_s_dist'] = df_5["s_dist"] - mean_s_dist
df_5.loc[:, 'vv_s_dist'] = df_5["v_s_dist"] ** 2
df_5

,dX,dY,dZ,s_dist,v_s_dist,vv_s_dist
DATE_TIME,,,,,,
2024-06-13 12:20:38,7760.790589,16649.359154,143.986395,18369.860176,-0.048893,2.390528e-03
2024-06-13 12:20:53,7760.796808,16649.406224,144.338888,18369.908231,-0.000838,7.023663e-07
2024-06-13 12:21:08,7760.867168,16649.409006,143.899801,18369.937033,0.027964,7.819959e-04
2024-06-13 12:21:23,7760.783906,16649.374532,143.526605,18369.867693,-0.041376,1.712003e-03
2024-06-13 12:21:38,7760.754749,16649.414683,144.019759,18369.895625,-0.013444,1.807488e-04
...,...,...,...,...,...,...
2024-06-13 12:47:23,7760.816609,16649.467900,144.274929,18369.971994,0.062925,3.959597e-03
2024-06-13 12:47:38,7760.867464,16649.429369,144.072238,18369.956966,0.047897,2.294133e-03
2024-06-13 12:48:08,7760.770582,16649.346013,144.451057,18369.843461,-0.065608,4.304355e-03


GnssVector(FMWY-WLTU, dist=18369.909, azimuth=65.008, zenith=89.550, mse_dx=0.058, mse_dy=0.060, mse_dz=0.297 mse_s_dist=0.006, mse_azimuth=0.069, mse_zenith=0.352)

In [57]:
mse_s_dist = ((df_5["vv_s_dist"].sum() / (len(df_5) - 1)) ** 0.5)
mse_s_dist

0.05703907736491693

In [37]:
mse_s_dist = ((df_5["vv_s_dist"].sum() / (len(df_5) - 1)) ** 0.5) / (len(df_5) ** 0.5)
mse_s_dist

0.006012446670256423

In [76]:
df_az = df_4[["dX", "dY", "dZ"]]
df_az['azimuth'] = np.degrees(np.arctan2(df_az['dY'], df_az['dX']))
df_az

/var/folders/c2/pfk652nd5r30mk7t_rz17pg00000gn/T/ipykernel_16488/3498326847.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_az['azimuth'] = np.degrees(np.arctan2(df_az['dY'], df_az['dX']))


,dX,dY,dZ,azimuth
DATE_TIME,,,,
2024-06-13 12:20:38,7760.790589,16649.359154,143.986395,65.008291
2024-06-13 12:20:53,7760.796808,16649.406224,144.338888,65.008336
2024-06-13 12:21:08,7760.867168,16649.409006,143.899801,65.008141
2024-06-13 12:21:23,7760.783906,16649.374532,143.526605,65.008331
2024-06-13 12:21:38,7760.754749,16649.414683,144.019759,65.008466
...,...,...,...,...
2024-06-13 12:47:23,7760.816609,16649.467900,144.274929,65.008361
2024-06-13 12:47:38,7760.867464,16649.429369,144.072238,65.008167
2024-06-13 12:48:08,7760.770582,16649.346013,144.451057,65.008331


In [77]:
mean_az = df_az["azimuth"].mean()
mean_az

65.00830051742676

In [78]:
df_az.loc[:, 'v_az'] = df_az["azimuth"] - mean_az
df_az.loc[:, 'vv_az'] = df_az["v_az"] ** 2
df_az

/var/folders/c2/pfk652nd5r30mk7t_rz17pg00000gn/T/ipykernel_16488/14610177.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_az.loc[:, 'v_az'] = df_az["azimuth"] - mean_az
/var/folders/c2/pfk652nd5r30mk7t_rz17pg00000gn/T/ipykernel_16488/14610177.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_az.loc[:, 'vv_az'] = df_az["v_az"] ** 2


,dX,dY,dZ,azimuth,v_az,vv_az
DATE_TIME,,,,,,
2024-06-13 12:20:38,7760.790589,16649.359154,143.986395,65.008291,-0.000009,8.206137e-11
2024-06-13 12:20:53,7760.796808,16649.406224,144.338888,65.008336,0.000035,1.252331e-09
2024-06-13 12:21:08,7760.867168,16649.409006,143.899801,65.008141,-0.000160,2.555471e-08
2024-06-13 12:21:23,7760.783906,16649.374532,143.526605,65.008331,0.000030,9.059313e-10
2024-06-13 12:21:38,7760.754749,16649.414683,144.019759,65.008466,0.000165,2.736930e-08
...,...,...,...,...,...,...
2024-06-13 12:47:23,7760.816609,16649.467900,144.274929,65.008361,0.000061,3.682405e-09
2024-06-13 12:47:38,7760.867464,16649.429369,144.072238,65.008167,-0.000134,1.791855e-08
2024-06-13 12:48:08,7760.770582,16649.346013,144.451057,65.008331,0.000030,9.110663e-10


In [80]:
mse_az = ((df_az["vv_az"].sum() / (len(df_az) - 1)) ** 0.5)
mse_az

0.00018926632760949975

In [81]:
mse_az = ((df_az["vv_az"].sum() / (len(df_az) - 1)) ** 0.5) / (len(df_az) ** 0.5)
mse_az

1.995042265405436e-05

In [82]:
mse_az * 3600

0.0718215215545957